### PR TITLE
feat : get products category

### PR DESCRIPTION
**Pull request by Amal-Salam**
###  What does the PR do?

     This PR addresses issue #257 by implementing the functionality to get product categories from the database.

### How should this be manually tested?
To test this feature, you can send a GET request to the following endpoint:
- [x] GET /api/v1/products/categories 

```json
    Authorization: Bearer <token>
```
     Successful Response

     On a successful products category retrieval, the API should return a 200  status code.

```json
{
  "status_code": 200,
  "categories": [
    {
      "id": "aee53410-ad20-4ac1-9876-cb650c75b88b",
      "created_at": "2024-07-24T20:08:05.581Z",
      "updated_at": "2024-07-24T20:08:05.581Z",
      "name": "Electronics",
      "description": "Electronic devices and accessories",
      "slug": "electronics",
      "parent_id": null
    }
```
success

Authentication Error:

When the current user is not properly authenticated to get the product category detail:

```json
  {
     "status": "error",
     "message": "User is currently unauthorized, kindly authenticate to continue",
     "status_code": 401
  }
```
no token image

Invalid Query Parameters

On Invalid error, if the query parameters are incorrect:

```json
{
  "error": {
    "code": "INVALID_QUERY_PARAMETER",
    "message": "The provided query parameter is invalid.",
    "details": {
      "invalid_parameter": "limit",
      "reason": "Must be a positive integer"
    }
  }
}
```
invalid query params

Validation:

If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages.

```json
{
  "message": "Cannot GET /api/v1/products/categorie",
  "error": "Not Found",
  "statusCode": 404
}
```
validation_error_image

Task
 - [x] Create the API endpoint for retrieving product categories.
 - [x] Implement efficient database querying to fetch category data.
 - [x] Include proper indexing on the categories table for optimal performance.
 - [x] Implement optional query parameters for advanced use cases (limit, offset).
 - [x] Ensure proper error handling for invalid requests or server-side issues.
 - [x] Write unit tests for the category retrieval logic.
 - [x] Update API documentation to reflect the new endpoint and its usage.

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/257)

### CODE REVIEW

File: src/modules/products/products.controller.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import { CreateProductDto } from './dto/create-product.dto';
+import { ApiTags, ApiBody } from '@nestjs/swagger';
+import { skipAuth } from '../../helpers/skipAuth';
+
+@Controller('products')
+export class ProductController {
+  constructor(private readonly productsService: ProductsService) {}
+
+  @skipAuth()
+  @ApiTags('Products')
```
Comment (davidolanrewaju) [inline src/modules/products/products.controller.ts:12]: Remove the @skipAuth() decorator

Comment (Amal-Salam) [reply src/modules/products/products.controller.ts:10]: Removed.thank you

**Comment by davidolanrewaju** [comment]
Your PR title should start with f not F

**Review by davidolanrewaju — APPROVED** [review]

**Review by davidolanrewaju — DISMISSED** [review]

File: .vscode/settings.json [hunk 2]
```diff
@@ -10,9 +10,14 @@
     "statusBar.background": "#1a1f15",
     "statusBarItem.hoverBackground": "#333d2a",
     "statusBar.foreground": "#e7e7e7",
-    "panel.border": "#333d2a",
-    "sideBar.border": "#333d2a",
-    "editorGroup.border": "#333d2a"
+    "commandCenter.border": "#e7e7e799",
```
Comment (Homoakin619) [inline .vscode/settings.json:13]: don't change the vscode settings, kindly revert this

File: src/modules/product-category/entities/product-category.entity.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Entity, PrimaryGeneratedColumn, Column, Index, OneToMany } from 'typeorm';
+import { Product } from '../../products/entities/product.entity';
+
+@Entity('categories')
+export class ProductCategory {
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
```
Comment (Homoakin619) [inline src/modules/product-category/entities/product-category.entity.ts:7]: remove this id column from here. Rather extend the AbstractBaseEntity it provides these columns out of the box

File: .vscode/settings.json (hunk 2 above)
Comment (Amal-Salam) [reply .vscode/settings.json]: noted with thanks

File: src/modules/product-category/dto/update-product-category.dto.ts
Comment (Heba-WebDev) [inline src/modules/product-category/dto/update-product-category.dto.ts:1]: This dto is empty, is it needed??

File: src/modules/product-category/entities/product-category.entity.ts
Comment (Heba-WebDev) [inline src/modules/product-category/entities/product-category.entity.ts:1]: Please specify the nullability of all columns, nullable: true/false. S

Comment (Amal-Salam) [reply src/modules/product-category/entities/product-category.entity.ts:1]: resolved

File: src/modules/product-category/dto/update-product-category.dto.ts
Comment (Amal-Salam) [reply src/modules/product-category/dto/update-product-category.dto.ts:1]: it is,i have made updates to it.
thank you

**Pull request closed without merging**